### PR TITLE
Migrate tool execution request from api/tools to api/jobs

### PR DESCRIPTION
**Pull request by guerler**
This PR switches tool execution to the canonical POST `/api/jobs` endpoint and removes the legacy `/api/tools` fallback from the frontend.

Before submission, the form state is transformed from the existing flat, pipe-separated structure into a nested RequestToolState that mirrors the backend model. All required type coercion is handled client-side so the request leaves the browser in a normalized, strictly typed form.

The backend remains strictly typed (strict: true by default) and required only small adjustments for feature parity, including support for tags, email notifications, object store preferences, and ID encoding.

After submission, the client polls `/api/tool_requests/{id}/state` until the request reaches a terminal state and then retrieves the outputs to render the success view.

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

File: client/src/components/Form/utilities.test.js [hunk 1]
```diff
@@ -524,4 +524,256 @@ describe("form component utilities", () => {
...
+                                    { value: "x", inputs: [{ name: "leaf", type: "text" }] },
+                                    { value: "y", inputs: [{ name: "other", type: "text" }] },
+                                ],
+                            },
+                        ],
+                    ],
+                    inputs: [],
+                },
+            ];
+            const formData = {
```
Comment (mvdbeek) [inline client/src/components/Form/utilities.test.js:627]: I'd love it if we did properly nested inputs here and dropped the legacy `|` syntax

Comment (guerler) [reply client/src/components/Form/utilities.test.js:627]: Yes, I agree, but this is about the form building side, which sits earlier in the flow. I did consider which piece to work on first and decided to switch the submission path to `/api/jobs` before refactoring the state building stage. The state handling improvements are a separate but equally important part of my form modernization roadmap. I will refactor the internal state to be properly nested and align it with the backend so the modernized form no longer needs to resubmit and re sync tool state.

File: test/unit/tool_util/parameter_specification.yml [hunk 2]
```diff
@@ -2297,26 +2296,24 @@ gx_data_column_optional:
...
   - { ref_parameter: {src: hda, id: 123}, parameter: 0 }
-  request_internal_invalid:
   - { ref_parameter: {src: hda, id: 123}, parameter: "0" }
 
 gx_data_column_multiple:
   request_valid:
   - { ref_parameter: {src: hda, id: abcdabcd}, parameter: [0] }
-  request_invalid:
   - { ref_parameter: {src: hda, id: abcdabcd}, parameter: "0" }
   - { ref_parameter: {src: hda, id: abcdabcd}, parameter: ["0"] }
```
Comment (jmchilton) [inline test/unit/tool_util/parameter_specification.yml:2351]: I'd like to keep these more strongly typed.

File: lib/galaxy/tool_util_models/parameters.py [hunk 3]
```diff
@@ -1679,11 +1695,23 @@ def pydantic_template(self, state_representation: StateRepresentationT) -> Dynam
...
+            if self.multiple:
+                validators = {
+                    "coerce_int_list": field_validator(self.name, mode="before")(
+                        DataColumnParameterModel.coerce_int_list
+                    )
+                }
+            else:
+                validators = {
+                    "coerce_int": field_validator(self.name, mode="before")(DataColumnParameterModel.coerce_int)
+                }
```
Comment (jmchilton) [inline lib/galaxy/tool_util_models/parameters.py]: Please do this client side if you're going to do it. Have you looked at my commit that attempts this transition? https://github.com/galaxyproject/galaxy/pull/17393/changes/7a9fb34a213d3cd2653f229ba8c8c3ea97c654ec? It pre-validates the the request before using /api/jobs - I think we want to do that even if you rewrite it from scratch. It also had one approach to use the meta-models on the client side - again - you can rewrite that approach also but I'd like to keep a strongly typed API and is validatable from the client.

There will be many cases where the current tool form produces less strongly typed requests - it might be worth just sticking a layer in there.

Comment (guerler) [reply lib/galaxy/tool_util_models/parameters.py]: Totally agreed, backend strict typing has to stay for this to become mergeable, any coercion should happen on the client before hitting `/api/jobs`. Thanks for referencing your commit.

Comment (jmchilton) [reply lib/galaxy/tool_util_models/parameters.py]: Great - thanks! I realize it will be an uphill battle to get everything working right but I think the stronger typing is already really helping with a bunch of use cases but I'll admit it does make the tool form piece a bit harder.

File: lib/galaxy/tool_util_models/parameters.py [hunk 4]
```diff
@@ -412,6 +412,11 @@ class DataRequestHdca(LegacyRequestModelAttributes):
     id: StrictStr
 
 
+class DataRequestDce(LegacyRequestModelAttributes):
+    src: Literal["dce"] = "dce"
+    id: StrictStr
```
Comment (jmchilton) [inline lib/galaxy/tool_util_models/parameters.py:470]: So far this stuff has pretty good API test coverage - can you add an API test for this in lib/galaxy_test/api/test_tool_execute.py? I guess this is for drag and dropping part of a collection?

File: lib/galaxy/webapps/galaxy/services/base.py [hunk 5]
```diff
@@ -201,25 +201,47 @@ def async_task_summary(async_result: AsyncResult) -> AsyncTaskResultSummary:
     )
 
 
-def tool_request_to_model(tool_request: ToolRequest) -> ToolRequestModel:
+def _encode_request_ids(request: dict[str, Any], security: IdEncodingHelper) -> dict[str, Any]:
```
Comment (jmchilton) [inline lib/galaxy/webapps/galaxy/services/base.py]: I don't want to walk the request and guess what is an object in there or not anymore - everything should be structured around the tool state and strongly typed. Code like this breaks when you have parameters that can be JSON. We have code to encode and decode requests in lib/galaxy/tool_util/parameters/convert.py - and many other operations like that. If those don't work - I think something like that should work. It will mean you'll have to pass the tool parameters into here but I think that is important for this.

File: lib/galaxy/managers/jobs.py [hunk 6]
```diff
@@ -2194,31 +2200,61 @@ def queue_jobs(self, tool: Tool, request: QueueJobs) -> None:
...
+                    )
+                for _, hdca in execution_tracker.output_collections:
+                    tag_handler.apply_item_tags(
+                        user=request_context.user, item=hdca, tags_str=",".join(request.tags), flush=False
+                    )
+            if request.send_email_notification:
+                if request_context.user is None:
+                    raise Exception("Anonymously run jobs cannot send an email notification.")
+                for job in execution_tracker.successful_jobs:
+                    job.add_post_job_action(PostJobAction("EmailAction"))
```
Comment (jmchilton) [inline lib/galaxy/managers/jobs.py]: These three operations seem essential and I love that you implemented them here - can you some how refactor more of this so it is code shared between this and the legacy path though? I guess these things shouldn't look that different.

File: lib/galaxy/schema/tasks.py [hunk 7]
```diff
@@ -186,3 +185,9 @@ class QueueJobs(Model):
     user: RequestUser  # TODO: test anonymous users through this submission path
     use_cached_jobs: bool
     rerun_remap_job_id: Optional[int]  # link to a job to rerun & remap
+    preferred_object_store_id: Optional[str] = None
+    tags: Optional[list[str]] = None
+    data_manager_mode: Optional[str] = None
+    send_email_notification: bool = False
+    credentials_context: Optional[list[dict]] = None
+    dynamic_tool_id: Optional[int] = None  # link to DynamicTool for custom/user tools
```
Comment (jmchilton) [inline lib/galaxy/schema/tasks.py:197]: Awesome - thank you for filling out these structures.

File: client/src/components/Tool/ToolForm.vue [hunk 8]
```diff
@@ -391,28 +399,26 @@ export default {
             }
             this.showExecuting = true;
             this.addRecentTool(this.formConfig?.id);
+
+            const nestedInputs = buildNestedState(this.formConfig.inputs, this.formData);
```
Comment (jmchilton) [inline client/src/components/Tool/ToolForm.vue:403]: I don't think all tools have a ``tool.parameters`` on the backend yet and by not using this on the frontend - I think you're going to break some tools. I hope I'm wrong - but I would do this with the fallback behavior for legacy tools.

**Comment by jmchilton** [comment]
It looks like I didn't implement subcollection mapping (map_over_type) on the backend for this API somehow. Maybe I'm misreading this but I looked over your stuff and through what is existing and I think what you added with map_over_type needs to be in there. I will take some time next week and walk through that and see if the map_over_type annotation makes sense as it is and write some API tests and model specification tests around it. Sorry for missing that during implementation.

**Comment by jmchilton** [comment]
I see you got it working without changing the existing interface and I really appreciated that - thank you so much. I'm not going to insist that you validate the client request on the frontend - but we have the schema (in multiple forms - including JSONschema) and we have the code - I just don't understand not doing that. The backend and the frontend are so complex - I would never just unleash this API without the validation and fallback to the legacy behavior - but I will let you do you and I will help adapt the API as we find issues.

File: lib/galaxy/tool_util_models/parameters.py (hunk 4 above)
Comment (guerler) [reply lib/galaxy/tool_util_models/parameters.py:470]: The client always resolves `dce` to `hda` before submission. The only time `dce` appears is for sub collection elements in batch or map over scenarios, which are now handled by collection expansion in `meta.py`.

**Comment by guerler** [comment]
@jmchilton Thanks for the review and for looking into the `map_over_type` gap. Honestly, I would prefer switching to the tool request API without a fallback and addressing issues as they come up. I added a log.warning for when tool.parameters fails to generate so affected tools remain visible. I intend to run the IUC tool test suite against this branch to gain more insight.

**Comment by mvdbeek** [comment]
Tool test are never going to trigger this. IWC maybe.

File: lib/galaxy_test/api/test_tool_execute.py [hunk 9]
```diff
@@ -470,20 +470,24 @@ def test_map_over_collection(
...
     """
     hdca = target_history.with_pair(["123", "456"])
     collection_details = target_history._dataset_populator.get_history_collection_details(
         target_history.id, content_id=hdca.id
     )
     forward_element = collection_details["elements"][0]
     assert forward_element["element_identifier"] == "forward"
-    dce_src_dict = {"src": "dce", "id": forward_element["id"]}
-    inputs = tool_input_format.when.any({"input1": dce_src_dict})
+    hda_id = forward_element["object"]["id"]
```
Comment (mvdbeek) [inline lib/galaxy_test/api/test_tool_execute.py]: Please don't do that. We should absolutely not do that, we lose a lot of traceability here. Submission via dce must work, that's a regression if you are doing this to fix a test.

File: lib/galaxy/celery/tasks.py [hunk 10]
```diff
@@ -363,7 +363,7 @@ def queue_jobs(request: QueueJobs, app: MinimalManagerApp, job_submitter: JobSub
     tool = cached_create_tool_from_representation(
         app=app,
         raw_tool_source=raw_tool_source,
-        tool_dir=request.tool_source.tool_dir,
+        tool_dir=request.tool_source.tool_dir or "",
```
Comment (mvdbeek) [inline lib/galaxy/celery/tasks.py]: Probably want to fix the type instead ?

File: lib/galaxy/managers/hdas.py [hunk 11]
```diff
@@ -182,6 +182,7 @@ def materialize(
             file_sources=self.app.file_sources,
             sa_session=session,
         )
+        assert request_user.user_id
```
Comment (mvdbeek) [inline lib/galaxy/managers/hdas.py:185]: If you're allowing anon async requests as you do in async_request_user you'll have to figure out what to do here.

File: lib/galaxy/managers/jobs.py [hunk 12]
```diff
@@ -2181,6 +2184,8 @@ def dereference_callback(data_request: DataRequestUri) -> DataRequestInternalHda
     def queue_jobs(self, tool: Tool, request: QueueJobs) -> None:
         tool_request: ToolRequest = self._tool_request(request.tool_request_id)
         sa_session = self.app.model.context
+        if request.dynamic_tool_id:
```
Comment (mvdbeek) [inline lib/galaxy/managers/jobs.py:2192]: Is the access validated on the other side of the request ? Not every dynamic tool is public.

File: lib/galaxy/schema/tasks.py (hunk 7 above)
Comment (mvdbeek) [reply lib/galaxy/schema/tasks.py:197]: There are no tests for this though, and without access checking this feels a little dangerous. I'd either add unprivileged and public access tests or drop this until the tests cover this

File: lib/galaxy/managers/jobs.py (hunk 12 above)
Comment (guerler) [reply lib/galaxy/managers/jobs.py:2192]: The access is validated in: https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/webapps/galaxy/services/tools.py#L103

Comment (mvdbeek) [reply lib/galaxy/managers/jobs.py:2192]:
```
    def allow_user_access(self, user, attempting_access=True):
        """
        :returns: bool -- Whether the user is allowed to access the tool.
        """
        if self.require_login and user is None:
            return False
        return True
```
So please fix or remove that.

Comment (mvdbeek) [reply lib/galaxy/managers/jobs.py:2192]: `get_tool_from_toolbox` probably does the access checks, but I'd love to see a separate test there.

Comment (guerler) [reply lib/galaxy/managers/jobs.py:2192]: Ok, thanks for catching this. I added an additional check to `allow_user_access`.

File: lib/galaxy/managers/hdas.py (hunk 11 above)
Comment (guerler) [reply lib/galaxy/managers/hdas.py:185]: This is a pre-existing issue, but I added an explicit check so anonymous users get a clear error instead of an assert failure.

File: lib/galaxy/tool_util_models/parameters.py (hunk 4 above)
Comment (jmchilton) [reply lib/galaxy/tool_util_models/parameters.py:470]: I'll include testing this in my work then - thanks for the clarification!

Comment (jmchilton) [reply lib/galaxy/tool_util_models/parameters.py:470]: Yeah - this fix seems needed I've placed it in https://github.com/galaxyproject/galaxy/pull/21991 with testing that should verify it is needed. GH Actions seem hosed for that branch though.

File: lib/galaxy/schema/tasks.py (hunk 7 above)
Comment (jmchilton) [reply lib/galaxy/schema/tasks.py:197]: I think this is fair - I'm happy to work on the tests if that would be helpful.

**Comment by jmchilton** [comment]
> Tool test are never going to trigger this. IWC maybe.

Depends on what we're trying to catch but tools without tool.parameters or request semantics that don't make sense also won't be caught with those. Running the tool tests with ``DEFAULT_USE_LEGACY_API=never`` set in the environment would I think potentially. If we had a clean idea of what works without the parameter set and breaks with it and we verified the test environment was using the new API - it would give us an idea of what we're not modeling currently.

Aside: The DCE thing is different and maybe would be caught with the workflow tests - I'm going to look into it though. Some initial tests are in https://github.com/galaxyproject/galaxy/pull/21991 - but that doesn't do DCE submission to tools - just DCEs in the DB after the request is expanded - I thought DCEs didn't work with the API but I'm going to research this point.

None of that verifies that "the blob" generated by the client is the right blob for the comparable request for what the old blob talking to the old API would be. Maybe wanting that level of correctness verification is too big an ask - but if we did want to do that - I think a framework for running all the framework tests as playwright tests through the browser would be the thing to do. It wouldn't be something we'd have to do all the time but we could setup the framework and get it working for all tests and then give this work a green checkmark in our hearts and then only run a few interesting corner cases with that framework during normal CI. This is a pattern we do use for job runners, embedded Pulsar, etc.. where we have the ability to run the whole suite but that is too much but we pick a few tools per obscure Galaxy config scenario. 

Skipping client side validation and not falling back when things go sideways - speaks to me of an urgency I don't understand. I would think getting the API solid and keeping the tool form working would be higher priorities than switching to the API at this pace. That said this remains my stated top priority - if agree that the tool tests should be all working without regression before merging this - I'll help implement the parameters and specification and testing around all the corner cases we find with the tool tests. Likewise - if tool form validation is something we agree should be there before this is merged - Claude and I can implement that Selenium framework and run it against all the tests in probably just week I want to say.

**Comment by guerler** [comment]
That makes sense. The goal of this PR was to surface issues early so we can fix them properly rather than discovering them late. I’m happy to take the time to get this right and still aim to get it into 26.1. For testing, I’m fine with whichever direction you think makes most sense. Thank you again for the work on the jobs API. It’s awesome.

**Comment by mvdbeek** [comment]
> Running the tool tests with `DEFAULT_USE_LEGACY_API=never` set in the environment would I think potentially.

There's no syntax for map over in tool tests so I don't think that can be exercised with existing IUC tests ? That's what my comment was about. It would of course be cool to see where we are with the new submission API, I can add a flag to planemo and enable celery tasks, are the tool framework tests already using this ? That'd be the first pass I guess

> Skipping client side validation and not falling back when things go sideways

I think this would be a useful thing beyond the migration period. The more we can validate on the frontend the better ?

**Comment by guerler** [comment]
@mvdbeek Agreed, client side validation would definitely be valuable. As discussed earlier around the nested inputs, the plan is to first stabilize the `/api/jobs` path with a solid backend validation, and then refactor the form generation api and state to be properly nested instead of the current flat-syntax. Once the form works with structured states that matches the API schema only, adding client side validation should become much cleaner.

**Review by jmchilton — APPROVED** [review]
Still uneasy for stated reasons - no client side validation, no ensuring the new endpoint works for tools, etc...

File: lib/galaxy/tools/__init__.py [hunk 13]
```diff
@@ -1321,6 +1321,19 @@ def allow_user_access(self, user, attempting_access=True):
...
+            try:
+                is_public = self.dynamic_tool.public
+            except Exception:
+                # DynamicTool not bound to session, access was validated at load time
+                is_public = True
+            if not is_public:
+                if user is None:
+                    return False
+                try:
+                    self.app.dynamic_tool_manager.ensure_can_use_unprivileged_tool(user)
```
Comment (mvdbeek) [inline lib/galaxy/tools/__init__.py:1334]: This just means the user has the USER_TOOL_EXECUTE role, it does not mean they can run **this** tool, that would be get_unprivileged_tool_by_uuid

File: test/unit/app/tools/test_toolbox.py [hunk 14]
```diff
@@ -156,6 +158,44 @@ def allow_user_access(user, attempting_access):
...
+
+        # Mock the dynamic tool manager to reject the user
+        mock_manager = MagicMock()
+        mock_manager.ensure_can_use_unprivileged_tool.side_effect = InsufficientPermissionsException(
+            "User is not allowed to run unprivileged tools"
+        )
+        tool.app.dynamic_tool_manager = mock_manager
+
+        # Non-public dynamic tool denies access without a user
+        assert tool.allow_user_access(None) is False
```
Comment (mvdbeek) [inline test/unit/app/tools/test_toolbox.py:182]: This tests none of the business logic

Comment (mvdbeek) [reply test/unit/app/tools/test_toolbox.py:182]: by using the mock manager all you're testing is that the manager method is called ...

File: lib/galaxy/tools/__init__.py (hunk 13 above)
Comment (guerler) [reply lib/galaxy/tools/__init__.py:1334]: Ugh, ok, switching this now.
